inc/AbstractPdoDbDriver.php

```diff
@@ -0,0 +1,713 @@
+	 *  - {@see PDO::FETCH_ASSOC} Fetch an array of results keyed by column name. This is the default.
+	 *  - {@see PDO::FETCH_NUM} Fetch an array of results keyed by column number, starting at 0.
+	 *  - {@see PDO::FETCH_BOTH} Fetch an array of results keyed by both column name and number.
+	 *
+	 * @return array|bool The array of results, or false if there are no more results.
+	 */
+	public function fetch_array($query, $resultType = PDO::FETCH_ASSOC)
+	{
+		if (is_null($query) || !($query instanceof PDOStatement)) {
+			return false;
+		}
+
+		switch($resultType)
+		{
+			case PDO::FETCH_NUM:
+			case PDO::FETCH_BOTH:
+				break;
+			default:
+				$resultType = PDO::FETCH_ASSOC;
+				break;
+		}
+
+		$hash = spl_object_hash($query);
+
+		if (isset($this->resultSeekPositions[$hash])) {
+			return $query->fetch($resultType, PDO::FETCH_ORI_ABS, $this->resultSeekPositions[$hash]);
+		}
+
+		return $query->fetch($resultType);
+	}
+
+	/**
+	 * Return a specific field from a query.
+	 *
+	 * @param PDOStatement $query The query to retrieve a result for.
+	 * @param string $field The name of the field to return.
+	 * @param int|bool $row The number of the row to fetch it from, or false to fetch from the next row in the result set.
+	 *
+	 * @return mixed The resulting field, of false if no more rows are in th result set.
+	 *  Note that when querying fields that have a boolean value, this method should not be used.
+	 */
+	public function fetch_field($query, $field, $row = false)
+	{
+		if (is_null($query) || !($query instanceof PDOStatement)) {
+			return false;
+		}
+
+		if ($row !== false) {
+			$this->data_seek($query, (int) $row);
+		}
+
+		// NOTE: PDOStatement::fetchColumn only operates on numbered columns, so we must fetch the array result
+		$array = $this->fetch_array($query, PDO::FETCH_ASSOC);
+
+		if ($array === false) {
+			return false;
+		}
+
+		return $array[$field];
+	}
+
+	/**
+	 * Move the internal row pointer to the specified row.
+	 *
+	 * @param PDOStatement $query The query to move the row pointer for.
+	 * @param int $row The row to move to. Rows are numbered from 0.
+	 *
+	 * @return bool Whether seeking was successful.
+	 */
+	public function data_seek($query, $row)
+	{
+		if (is_null($query) || !($query instanceof PDOStatement)) {
+			return false;
+		}
+
+		$hash = spl_object_hash($query);
+
+		// NOTE: PDO numbers rows from 1, but all other drivers are 0 based. We add 1 to the row number for compatibility
+		$this->resultSeekPositions[$hash] = ((int) $row) + 1;
+
+		return true;
+	}
+
+	/**
+	 * Return the number of rows resulting from a query.
+	 *
+	 * @param PDOStatement $query The query data.
+	 * @return int|bool The number of rows in the result, or false on failure.
+	 */
+	public function num_rows($query)
+	{
+		if (is_null($query) || !($query instanceof PDOStatement)) {
+			return false;
+		}
+
+		if (preg_match('/^\\s*SELECT\\b/i', $query->queryString) === 1) {
```

Comment:
Note: using `num_rows` with PDO is not optimal at all. We need to review usages within the core and see when/why we use `num-rows` and if they can be replaced so that we don't hit this path too often.

This works better with the old `mysql_` type functions as those functions buffer results internally, which PDO does not do.